### DISCUSSION TITLE
PHPMailer not following HTML special chars

### DISCUSSION THREAD

**Discussion by jweaver0312** [PHPMailer/PHPMailer#2724] [Q&A]
For example, if I use `&dollar;` while PHPMailer is set to HTML content, which would translate in HTML to just the $ sign, PHPMailer sends it as `&dollar;` in plaintext and not translating it to the '$'. If I explicitly have it send a '$', it works fine.

Is there a way to get it to translate the `&dollar;` to '$'?

As another example, `&nbsp;` works fine.

**Comment by Synchro** [comment]
PHPMailer doesn't care what content you use; it's entirely up to you how to handle it. Plain-text messages do not know about HTML entities, so they won't work in that context (presumably what you're seeing). If you want to convert HTML entities back to real characters, pass your message body through [`html_entity_decode`](https://www.php.net/manual/en/function.html-entity-decode.php) first, though make sure that your `CharSet` actually contains the character you want (i.e. if it's anything other than UTF-8).

**Reply by jweaver0312** [reply]
it is set as a HTML via `$mail->isHTML(true);`

Shouldn't it already be acknowledging `&dollar;` as '$' in that context? Since it is set to HTML message, shouldn't it be acknowledging HTML entities?

**Reply by Synchro** [reply]
Yes - make sure that you're not double-encoding. Look at the raw received message to check what's actually there.

**Reply by jweaver0312** [reply]
Ran it through `html_entity_decode()` and it now gave the proper result. Any idea as towards why `&amp;` and `&nbsp;` were recognized without me having to run it through `html_entity_decode()`, but then `&dollar;` wasn't?

**Reply by Synchro** [reply]
The most likely reason is double encoding (e.g. `&amp;dollar;`, decodes once to `&dollar;`, then a second time to `$`)) sneaking into your content somehow. Double-check your raw delivered message to be sure. It could also be that some clients don't decode `&dollar;` correctly, since it's supported in ASCII anyway?

**Reply by jweaver0312** [reply]
The test I have just has `&dollar;` in the body and only that but then it still just comes out as '&dollar;' where the only time I get `&dollar;` correctly to come back as '$' is when ran through `html_entity_decode('&dollar;', ENT_HTML5)` but if I'm doing something else general with PHP, the `&dollar;` is correctly decoded without the need of `html_entity_decode()`

**Reply by Synchro** [reply]
It would help if you could show the raw message that doesn't work as expected. Bear in mind that `&dollar;` *is* expected to appear in HTML source code, but it will be rendered as `$` when viewed.

**Reply by jweaver0312** [reply]
Just so that we're on the same page, by raw message, you do mean what I set body to, correct?

If so, I did 2 tests where the raw content was `&dollar;`.

One was this (where I wrote a mostly full HTML format):
`<!doctype html>
<head>
<title>test</title>
</head>
<body>
&dollar;
</body>
</html>`

and the 2nd test I did was just setting body to be `&dollar;`

Both of those are while setting `isHTML(true);`

**Reply by Synchro** [reply]
No, I mean a *received* message, so we can see exactly what was delivered. You can get this in gmail and other clients via options like "show original". You can also see the content you are sending if you set `SMTPDebug = 2`. The idea is to compare what you put in to `Body` versus what was sent.
That example you gave is not even remotely valid HTML, so don't be surprised when things don't look right.

**Reply by jweaver0312** [reply]
I had more in the HTML it sent, but I didn't format it correctly for Github to render it correctly (as in I have head and body tags as well in that.)

Went through it again, though I also temporarily removed base64 encoding to be able to see the raw HTML and the following is received

```
<!doctype html>
<html>
<head>
<title>Test</title>
</head>
<body>
This is a test email in HTML.
<br><br>
Testing some HTML entities:&nbsp;&dollar; (dollar sign expected)&nbsp;&amp (ampersand expected);
<br><br>
<b>This is in bold</b>
<br><br>
This word is in&nbsp;<s>strikethrough</s>
</body>
</html>
```

**Reply by Synchro** [reply]
OK, that's a reasonable test, and it is correctly formatted, apart from the `&amp` missing a trailing `;`.

Why are you using base64 encoding? It will get your messages classed as spam fairly reliably, there's really no good reason to do that.

**Reply by jweaver0312** [reply]
Only reason I did base64, is because without me doing it, it returns `Content-Type: text/html; charset=us-ascii` when I have CharSet set to 'UTF-8'.

I was revising the test script on the fly, but other than that this is what actually renders (after I moved the semicolon back to where it should be.)

This is a test email in HTML. 

Testing some HTML entities: \&dollar; (dollar sign expected) & (ampersand expected) 

**This is in bold** 

This word is in ~~strikethrough~~

**Reply by Synchro** [reply]
PHPMailer changes the charset automatically because sending email specifying an 8-bit charset when you're only using 7-bit chars is an RFC contravention. Put in a character that requires an 8-bit encoding and it won't change to ASCII. If those chars are encoded identically, the difference will be down to the client. The dollar sign is not special in HTML, so there is no particular reason to encode it as an entity. I also did another test that confirmed that any entity that is not understood by the client will be rendered as is, without conversion to a character. Try `&abcxyz;` and you'll see it remains untouched.

**Reply by jweaver0312** [reply]
`&abcxyz;` remains untouched and is rendered as is without any conversion to a character.

By saying it is not understood by the client, you are referring to the browser itself, correct? If I copy the raw contents of the message and put it into an html file and then view it on the browser, the `&dollar;` is then rendered as $. Only reason I asked about the `&dollar;` is because it would retrieve some data from the database that was put in with a different editor that made $ to be `&dollar;` before I switched to a new editor.

If I got to, I'll run it through `html_entity_decode()` or I'll go in updating the `&dollar;`s to be actual dollar signs.
